src/OpenTelemetry.Exporter.Console/ConsoleMetricExporter.cs

```diff
@@ -32,7 +32,7 @@ public ConsoleMetricExporter(ConsoleExporterOptions options)
         {
         }
 
-        public override ExportResult Export(in Batch<MetricItem> batch)
+        public override ExportResult Export(in Batch<Metric> batch)
```

Comment:
I think 1) metrics exporter will always run in a separate thread (unlike trace/log exporter which might run in the hot path - e.g. SimpleExportingProcessor) 2) metrics exporter will only be triggered at a much lower frequency (unlike trace/log exporter which might get triggered for every single event), so it might be actually easier if we define this as `IEnumerable<Metric>`.

I understand that there might be a desire to have `ConsoleExporter<T>` which covers logs/metrics/traces (for sake of consistency).

@CodeBlanch do you have strong opinion on this? (look at the change on the `src/OpenTelemetry/Batch.cs`, do you think the extra if-condition could slow down traces/logs?)

Reply:
Yes agree to make this simpler with just `IEnumerable`, so we won't affect the other signals's hot path.

Reply:
Could we do IList<T> instead of IEnumerable<T>? With that we could at least do a `for (int i = 0; i < list.Count; i++) { var item = list[i]; }` type of loop that  doesn't allocate an enumerator. Or we could even pass the array directly. Array doesn't have a struct enumerator but AFAIK the JIT is smart enough to re-write a foreach loop on an array to be allocation-free.

Reply:
I don't know if we will be able to tell the count of metrics or not, if we could, IList definitely works better.

I _guess_ array won't work since it assumes certain memory layout. For example, if we pre-allocate memory and keep reusing them, we would need to compact the pre-allocated aggregation buffer before exporting (and that seems to be expensive).

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet/pull/2306/files#diff-4d53a254dbe35c9d4bcfd7b6523549ab07b5372f3493d79b76797c724720cf75R170

I added a comment about this in the follow up PR.

Reply:
We could also pass a ReadOnlySpan<T> over the array?